File: src/lib/entry/command/command-type/json-command-type.ts
import CommandLogEntry from "../../command-log-entry"

export type JSONCommandTypeArgs = {
    commandNameU8?: Uint8Array,
    commandValueU8?: Uint8Array,
    value?: any,
}

export default class JSONCommandType extends CommandLogEntry {
    constructor(args: JSONCommandTypeArgs) {
        if (args.commandNameU8 && args.commandValueU8) {
            super({ commandNameU8: args.commandNameU8, commandValueU8: args.commandValueU8 })
        }
        else if (args.commandNameU8 && args.value !== undefined) {
            if (typeof args.value !== "string") {
                args.value = JSON.stringify(args.value)
            }
            super({
                commandNameU8: args.commandNameU8,
                commandValueU8: new TextEncoder().encode(args.value),
            })
        }
        else {
            throw new Error("JSONCommandType requires commandNameU8 and either commandValueU8 or value")
        }
    }

    value(): any {
        const text = new TextDecoder().decode(this.commandValueU8)
        return JSON.parse(text)
    }
}

File: external/logsrdcpp/src/entry/command/JSONCommandType.cpp
#include "JSONCommandType.h"
#include <cstring>

namespace logsrd {

JSONCommandType::JSONCommandType(JSONCommandTypeArgs args)
    : CommandLogEntry(
        std::move(args.commandNameU8),
        args.hasValue
            ? std::vector<uint8_t>(args.value.begin(), args.value.end())
            : std::move(args.commandValueU8))
{}

std::string JSONCommandType::value() const {
    return std::string(commandValueU8.begin(), commandValueU8.end());
}

void JSONCommandType::setValue(const std::string& val) {
    commandValueU8.assign(val.begin(), val.end());
}

} // namespace logsrd
